multiplex-rs/src/lib.rs

```diff
@@ -0,0 +1,617 @@
+            meta: muxer.meta,
+            state: muxer.state,
+        }
+    }
+}
+
+fn nonce_to_id(id: usize, end: Endpoint) -> u32 {
+    id as u32 * 2 + if end == Endpoint::Dialer { 0 } else { 1 }
+}
+
+impl<T: AsyncWrite> Future for OutboundFuture<T> {
+    type Item = Substream<T>;
+    type Error = io::Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        let mut lock = match self.state.poll_lock() {
+            Async::Ready(lock) => lock,
+            Async::NotReady => return Ok(Async::NotReady),
+        };
+
+        loop {
+            let (mut id_str, id) = self.current_id.take().unwrap_or_else(|| {
+                let next = nonce_to_id(
+                    self.meta.nonce.fetch_add(1, atomic::Ordering::Relaxed),
+                    self.meta.end,
+                );
+                (
+                    io::Cursor::new(buf_from_slice(format!("{}", next).as_bytes())),
+                    next as u32,
+                )
+            });
+
+            match write_stream(
+                &mut *lock,
+                write::WriteRequest::meta(MultiplexHeader::open(id)),
+                &mut id_str,
+            ) {
+                Ok(_) => {
+                    debug_assert!(id_str.position() <= id_str.get_ref().len() as u64);
+                    if id_str.position() == id_str.get_ref().len() as u64 {
+                        if lock.open_stream(id) {
+                            return Ok(Async::Ready(Substream::new(
+                                id,
+                                self.meta.end,
+                                Bytes::from(&id_str.get_ref()[..]),
+                                Arc::clone(&self.state),
+                            )));
+                        }
+                    } else {
+                        self.current_id = Some((id_str, id));
+                        return Ok(Async::NotReady);
+                    }
+                }
+                Err(ref err) if err.kind() == io::ErrorKind::WouldBlock => {
+                    self.current_id = Some((id_str, id));
+
+                    return Ok(Async::NotReady);
+                }
+                Err(other) => return Err(other),
+            }
+        }
+    }
+}
+
+pub struct MultiplexMetadata {
+    nonce: AtomicUsize,
+    end: Endpoint,
+}
+
+pub struct Multiplex<T> {
+    meta: Arc<MultiplexMetadata>,
+    state: Arc<Mutex<MultiplexShared<T>>>,
+}
+
+impl<T> Clone for Multiplex<T> {
+    fn clone(&self) -> Self {
+        Multiplex {
+            meta: self.meta.clone(),
+            state: self.state.clone(),
+        }
+    }
+}
+
+impl<T> Multiplex<T> {
+    pub fn new(stream: T, end: Endpoint) -> Self {
+        Multiplex {
+            meta: Arc::new(MultiplexMetadata {
+                nonce: AtomicUsize::new(0),
+                end,
+            }),
+            state: Arc::new(Mutex::new(MultiplexShared::new(stream))),
+        }
+    }
+
+    pub fn dial(stream: T) -> Self {
```

Comment:
I don't think this function and `listen` (just below) are necessary?

Reply:
I use them for tests and they're useful to manually construct a `Multiplex` if you're using it without `libp2p`